// mdb-angular-wysiwyg/esm2020/lib/wysiwyg.module.mjs
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { MdbWysiwygComponent } from './wysiwyg.component';
import { MdbDropdownModule } from 'mdb-angular-ui-kit/dropdown';
import { MdbFormsModule } from 'mdb-angular-ui-kit/forms';
import { MdbTooltipModule } from 'mdb-angular-ui-kit/tooltip';
import * as i0 from "@angular/core";
export class MdbWysiwygModule {
}
MdbWysiwygModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbWysiwygModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbWysiwygModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbWysiwygModule, declarations: [MdbWysiwygComponent], imports: [ReactiveFormsModule, CommonModule, MdbDropdownModule, MdbFormsModule, MdbTooltipModule], exports: [MdbWysiwygComponent] });
MdbWysiwygModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbWysiwygModule, imports: [[ReactiveFormsModule, CommonModule, MdbDropdownModule, MdbFormsModule, MdbTooltipModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbWysiwygModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbWysiwygComponent],
                    imports: [ReactiveFormsModule, CommonModule, MdbDropdownModule, MdbFormsModule, MdbTooltipModule],
                    exports: [MdbWysiwygComponent],
                }]
        }] });
