// mdb-angular-drag-and-drop/esm2020/lib/drag-and-drop.module.mjs
import { NgModule } from '@angular/core';
import { MdbDraggableDirective } from './draggable.directive';
import { MdbSortableContainerDirective } from './sortable-container.directive';
import * as i0 from "@angular/core";
export class MdbDragAndDropModule {
}
MdbDragAndDropModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbDragAndDropModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbDragAndDropModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbDragAndDropModule, declarations: [MdbDraggableDirective, MdbSortableContainerDirective], exports: [MdbDraggableDirective, MdbSortableContainerDirective] });
MdbDragAndDropModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbDragAndDropModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbDragAndDropModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbDraggableDirective, MdbSortableContainerDirective],
                    imports: [],
                    exports: [MdbDraggableDirective, MdbSortableContainerDirective],
                }]
        }] });
