// mdb-angular-drag-and-drop/esm2020/lib/utilities.mjs
export function moveItemToNewContainer(previousContainer, newContainer, previousIndex, newIndex) {
    if (previousContainer.length > 0) {
        const item = previousContainer.splice(previousIndex, 1)[0];
        newContainer.splice(newIndex, 0, item);
    }
}
export function copyItemToNewContainer(previousContainer, newContainer, previousIndex, newIndex) {
    if (previousContainer.length > 0) {
        const item = previousContainer[previousIndex];
        newContainer.splice(newIndex, 0, item);
    }
}
export function moveItemsInContainer(container, previousIndex, newIndex) {
    const item = container[previousIndex];
    container.splice(previousIndex, 1);
    container.splice(newIndex, 0, item);
}
export function getElementRect(element) {
    const elementRect = element.getBoundingClientRect();
    return {
        top: elementRect.top,
        right: elementRect.right,
        bottom: elementRect.bottom,
        left: elementRect.left,
        width: elementRect.width,
        height: elementRect.height,
    };
}
export function cloneNode(node) {
    const clone = node.cloneNode(true);
    clone.removeAttribute('id');
    return clone;
}
export function destroyNode(node) {
    if (node && node.parentNode) {
        node.parentNode.removeChild(node);
    }
}
