// mdb-angular-transfer/esm2020/lib/transfer.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { MdbTransferComponent } from './transfer.component';
import { MdbTransferContainerComponent } from './transfer-container.component';
import { MdbFormsModule } from 'mdb-angular-ui-kit/forms';
import * as i0 from "@angular/core";
export class MdbTransferModule {
}
MdbTransferModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbTransferModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbTransferModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbTransferModule, declarations: [MdbTransferComponent, MdbTransferContainerComponent], imports: [CommonModule, FormsModule, MdbFormsModule], exports: [MdbTransferComponent] });
MdbTransferModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbTransferModule, imports: [[CommonModule, FormsModule, MdbFormsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbTransferModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbTransferComponent, MdbTransferContainerComponent],
                    imports: [CommonModule, FormsModule, MdbFormsModule],
                    exports: [MdbTransferComponent],
                }]
        }] });
