// mdb-angular-ui-kit/esm2020/sidenav/sidenav-item.component.mjs
import { ChangeDetectionStrategy, Component, ContentChild, forwardRef, Inject, Optional, ViewChild, } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { MdbCollapseDirective } from 'mdb-angular-ui-kit/collapse';
import { MdbSidenavComponent } from './sidenav.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./sidenav.component";
export class MdbSidenavItemComponent {
    constructor(_renderer, _elRef, _router, _route, _cdRef, _sidenav) {
        this._renderer = _renderer;
        this._elRef = _elRef;
        this._router = _router;
        this._route = _route;
        this._cdRef = _cdRef;
        this._sidenav = _sidenav;
        this._tempSlim = false;
        this._isSlimTransitioning = false;
        this._destroy$ = new Subject();
    }
    ngAfterViewInit() {
        if (!this.collapse) {
            return;
        }
        const links = this._elRef.nativeElement.querySelectorAll('.sidenav-link');
        links[0].innerHTML += '<i class="fas fa-angle-down rotate-icon"></i>';
        if (!this.collapse.collapsed) {
            this._toggleArrowIcon('up');
        }
        this._sidenav.sidenavCollapsed.pipe(takeUntil(this._destroy$)).subscribe(() => {
            this._isSlimTransitioning = false;
        });
        this._sidenav.sidenavCollapse.pipe(takeUntil(this._destroy$)).subscribe(() => {
            this._isSlimTransitioning = true;
        });
        this._sidenav.sidenavExpand.pipe(takeUntil(this._destroy$)).subscribe(() => {
            this._isSlimTransitioning = true;
        });
        this._sidenav.sidenavExpanded.pipe(takeUntil(this._destroy$)).subscribe(() => {
            this._isSlimTransitioning = false;
        });
        this.collapse.collapseHide.pipe(takeUntil(this._destroy$)).subscribe(() => {
            if (!this._sidenav.slim) {
                return;
            }
            if (this._sidenav.isTheLastItemToBeCollapsed() &&
                !this._sidenav.slimCollapsed &&
                !this._isSlimTransitioning) {
                this._tempSlim = false;
                this._sidenav.setSlim(true);
            }
        });
        this.collapse.host.childNodes.forEach((child) => {
            fromEvent(child, 'click')
                .pipe(takeUntil(this._destroy$))
                .subscribe(() => this.setActiveElement(child));
        });
        fromEvent(this._sidenavLink.nativeElement, 'click')
            .pipe(takeUntil(this._destroy$))
            .subscribe(() => {
            if (this._sidenav.accordion) {
                this._sidenav.collapseItems();
            }
            this.collapse.toggle();
            const arrowDirection = this.collapse.collapsed ? 'down' : 'up';
            this._toggleArrowIcon(arrowDirection);
            if (this._sidenav.slimCollapsed && !this._isSlimTransitioning) {
                this._tempSlim = true;
                this._sidenav.setSlim(false);
            }
        });
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
    markForCheck() {
        this._cdRef.markForCheck();
    }
    setActiveElement(el) {
        this.markForCheck();
        this.collapse.host.childNodes.forEach((child) => {
            if (child instanceof HTMLElement) {
                if (el === child) {
                    this._renderer.addClass(child.querySelector('.sidenav-link'), 'active');
                }
                else {
                    this._renderer.removeClass(child.querySelector('.sidenav-link'), 'active');
                }
            }
        });
    }
    setActiveCategory() {
        this.markForCheck();
        this._renderer.addClass(this._sidenavLink.nativeElement.querySelector('.sidenav-link'), 'active');
    }
    unsetActiveCategory() {
        this.markForCheck();
        this._renderer.removeClass(this._sidenavLink.nativeElement.querySelector('.sidenav-link'), 'active');
    }
    _toggleArrowIcon(direction) {
        const arrowIcon = this._sidenavLink.nativeElement.querySelector('.fa-angle-down');
        if (direction === 'down') {
            this._renderer.setStyle(arrowIcon, 'transform', 'rotate(0deg)');
        }
        else {
            this._renderer.setStyle(arrowIcon, 'transform', 'rotate(180deg)');
        }
    }
}
MdbSidenavItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbSidenavItemComponent, deps: [{ token: i0.Renderer2 }, { token: i0.ElementRef }, { token: i1.Router, optional: true }, { token: i1.ActivatedRoute, optional: true }, { token: i0.ChangeDetectorRef }, { token: forwardRef(() => MdbSidenavComponent) }], target: i0.ɵɵFactoryTarget.Component });
MdbSidenavItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.7", type: MdbSidenavItemComponent, selector: "mdb-sidenav-item", queries: [{ propertyName: "collapse", first: true, predicate: MdbCollapseDirective, descendants: true }], viewQueries: [{ propertyName: "_sidenavLink", first: true, predicate: ["linkWrapper"], descendants: true }], ngImport: i0, template: "<li class=\"sidenav-item\">\n  <div #linkWrapper>\n    <ng-content select=\".sidenav-link\"></ng-content>\n  </div>\n  <ng-content select=\".sidenav-collapse\"></ng-content>\n</li>\n", changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbSidenavItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'mdb-sidenav-item', changeDetection: ChangeDetectionStrategy.OnPush, template: "<li class=\"sidenav-item\">\n  <div #linkWrapper>\n    <ng-content select=\".sidenav-link\"></ng-content>\n  </div>\n  <ng-content select=\".sidenav-collapse\"></ng-content>\n</li>\n" }]
        }], ctorParameters: function () { return [{ type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i1.Router, decorators: [{
                    type: Optional
                }] }, { type: i1.ActivatedRoute, decorators: [{
                    type: Optional
                }] }, { type: i0.ChangeDetectorRef }, { type: i2.MdbSidenavComponent, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => MdbSidenavComponent)]
                }] }]; }, propDecorators: { _sidenavLink: [{
                type: ViewChild,
                args: ['linkWrapper']
            }], collapse: [{
                type: ContentChild,
                args: [MdbCollapseDirective]
            }] } });
