// mdb-angular-vector-maps/esm2020/lib/vector-map.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MdbTooltipModule } from 'mdb-angular-ui-kit/tooltip';
import { MdbVectorMapComponent } from './vector-map.component';
import * as i0 from "@angular/core";
export class MdbVectorMapModule {
}
MdbVectorMapModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbVectorMapModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbVectorMapModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbVectorMapModule, declarations: [MdbVectorMapComponent], imports: [CommonModule, MdbTooltipModule, OverlayModule], exports: [MdbVectorMapComponent] });
MdbVectorMapModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbVectorMapModule, imports: [[CommonModule, MdbTooltipModule, OverlayModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbVectorMapModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbVectorMapComponent],
                    imports: [CommonModule, MdbTooltipModule, OverlayModule],
                    exports: [MdbVectorMapComponent],
                }]
        }] });
