// mdb-angular-file-upload/esm2020/lib/file-upload.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MdbFileUploadComponent } from './file-upload.component';
import * as i0 from "@angular/core";
export class MdbFileUploadModule {
}
MdbFileUploadModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbFileUploadModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbFileUploadModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbFileUploadModule, declarations: [MdbFileUploadComponent], imports: [CommonModule], exports: [MdbFileUploadComponent] });
MdbFileUploadModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbFileUploadModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbFileUploadModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbFileUploadComponent],
                    imports: [CommonModule],
                    exports: [MdbFileUploadComponent],
                }]
        }] });
