// mdb-angular-calendar/esm2020/lib/calendar.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { MdbCalendarToolsComponent } from './calendar-tools.component';
import { MdbCalendarComponent } from './calendar.component';
import { MdbCalendarMonthViewComponent } from './calendar-month-view.component';
import { MdbCalendarWeekViewComponent } from './calendar-week-view.component';
import { MdbCalendarListViewComponent } from './calendar-list-view.component';
import { MdbCalendarEventModalComponent } from './calendar-event-modal.component';
import { MdbFormsModule } from 'mdb-angular-ui-kit/forms';
import { MdbDatepickerModule } from 'mdb-angular-ui-kit/datepicker';
import { MdbTimepickerModule } from 'mdb-angular-ui-kit/timepicker';
import { MdbTooltipModule } from 'mdb-angular-ui-kit/tooltip';
import { MdbModalModule } from 'mdb-angular-ui-kit/modal';
import { MdbValidationModule } from 'mdb-angular-ui-kit/validation';
import * as i0 from "@angular/core";
export class MdbCalendarModule {
}
MdbCalendarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbCalendarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbCalendarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbCalendarModule, declarations: [MdbCalendarComponent,
        MdbCalendarToolsComponent,
        MdbCalendarMonthViewComponent,
        MdbCalendarWeekViewComponent,
        MdbCalendarListViewComponent,
        MdbCalendarEventModalComponent], imports: [CommonModule,
        MdbFormsModule,
        MdbDatepickerModule,
        MdbTimepickerModule,
        MdbTooltipModule,
        ReactiveFormsModule,
        MdbModalModule,
        MdbValidationModule], exports: [MdbCalendarComponent,
        MdbCalendarToolsComponent,
        MdbCalendarMonthViewComponent,
        MdbCalendarWeekViewComponent,
        MdbCalendarListViewComponent,
        MdbCalendarEventModalComponent] });
MdbCalendarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbCalendarModule, imports: [[
            CommonModule,
            MdbFormsModule,
            MdbDatepickerModule,
            MdbTimepickerModule,
            MdbTooltipModule,
            ReactiveFormsModule,
            MdbModalModule,
            MdbValidationModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbCalendarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MdbCalendarComponent,
                        MdbCalendarToolsComponent,
                        MdbCalendarMonthViewComponent,
                        MdbCalendarWeekViewComponent,
                        MdbCalendarListViewComponent,
                        MdbCalendarEventModalComponent,
                    ],
                    imports: [
                        CommonModule,
                        MdbFormsModule,
                        MdbDatepickerModule,
                        MdbTimepickerModule,
                        MdbTooltipModule,
                        ReactiveFormsModule,
                        MdbModalModule,
                        MdbValidationModule,
                    ],
                    exports: [
                        MdbCalendarComponent,
                        MdbCalendarToolsComponent,
                        MdbCalendarMonthViewComponent,
                        MdbCalendarWeekViewComponent,
                        MdbCalendarListViewComponent,
                        MdbCalendarEventModalComponent,
                    ],
                }]
        }] });
