// mdb-angular-calendar/esm2020/lib/calendar-tools.component.mjs
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { addDays, getDate, getDayNumber, getMonth, getYear } from './calendar.utils';
import * as i0 from "@angular/core";
export class MdbCalendarToolsComponent {
    constructor() {
        this.viewChange = new EventEmitter();
        this.previousBtnClick = new EventEmitter();
        this.nextBtnClick = new EventEmitter();
        this.todayBtnClick = new EventEmitter();
    }
    get activeDate() {
        return this._activeDate;
    }
    set activeDate(date) {
        this._activeDate = date;
        this.setPeriod();
    }
    get activeView() {
        return this._activeView;
    }
    set activeView(view) {
        this._activeView = view;
        this.setPeriod();
    }
    setPeriod() {
        switch (this.activeView) {
            case 'month':
                this.period = `${this.options.months[this.activeDate.getMonth()]} ${this.activeDate.getFullYear()}`;
                break;
            default:
                const sundayIndex = this.mondayFirst ? 1 : 0;
                const firstDay = addDays(this.activeDate, -getDayNumber(this.activeDate) + sundayIndex);
                const lastDay = addDays(firstDay, 6);
                const periodStart = `${this.options.monthsShort[getMonth(firstDay)]} ${getDate(firstDay)}`;
                const periodEnd = `${this.options.monthsShort[getMonth(lastDay)]} ${getDate(lastDay)}, ${getYear(lastDay)}`;
                this.period = `${periodStart} - ${periodEnd}`;
                break;
        }
    }
    toggleView(view) {
        this.viewChange.emit(view);
    }
    handlePreviousBtnClick() {
        this.previousBtnClick.emit();
    }
    handleNextBtnClick() {
        this.nextBtnClick.emit();
    }
    handleTodayBtnClick() {
        this.todayBtnClick.emit();
    }
}
MdbCalendarToolsComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbCalendarToolsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
MdbCalendarToolsComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.3", type: MdbCalendarToolsComponent, selector: "mdb-tools-calendar", inputs: { options: "options", activeDate: "activeDate", mondayFirst: "mondayFirst", activeView: "activeView" }, outputs: { viewChange: "viewChange", previousBtnClick: "previousBtnClick", nextBtnClick: "nextBtnClick", todayBtnClick: "todayBtnClick" }, ngImport: i0, template: "<div class=\"calendar-tools\">\n  <div class=\"btn-group btn-group-sm shadow-0\">\n    <button class=\"btn btn-outline-primary\" (click)=\"handlePreviousBtnClick()\">\n      <i class=\"fas fa-chevron-left\"></i>\n    </button>\n    <button class=\"btn btn-outline-primary\" (click)=\"handleNextBtnClick()\">\n      <i class=\"fas fa-chevron-right\"></i>\n    </button>\n    <button class=\"btn btn-outline-primary\" (click)=\"handleTodayBtnClick()\">{{ options.todayCaption }}</button>\n  </div>\n  <span class=\"calendar-heading\">{{ period }}</span>\n  <div class=\"btn-group btn-group-sm shadow-0\">\n    <button class=\"btn btn-outline-primary\" (click)=\"toggleView('month')\">{{ options.monthCaption }}</button>\n    <button class=\"btn btn-outline-primary\" (click)=\"toggleView('week')\">{{ options.weekCaption }}</button>\n    <button class=\"btn btn-outline-primary\" (click)=\"toggleView('list')\">{{ options.listCaption }}</button>\n  </div>\n</div>" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbCalendarToolsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'mdb-tools-calendar', styles: [], template: "<div class=\"calendar-tools\">\n  <div class=\"btn-group btn-group-sm shadow-0\">\n    <button class=\"btn btn-outline-primary\" (click)=\"handlePreviousBtnClick()\">\n      <i class=\"fas fa-chevron-left\"></i>\n    </button>\n    <button class=\"btn btn-outline-primary\" (click)=\"handleNextBtnClick()\">\n      <i class=\"fas fa-chevron-right\"></i>\n    </button>\n    <button class=\"btn btn-outline-primary\" (click)=\"handleTodayBtnClick()\">{{ options.todayCaption }}</button>\n  </div>\n  <span class=\"calendar-heading\">{{ period }}</span>\n  <div class=\"btn-group btn-group-sm shadow-0\">\n    <button class=\"btn btn-outline-primary\" (click)=\"toggleView('month')\">{{ options.monthCaption }}</button>\n    <button class=\"btn btn-outline-primary\" (click)=\"toggleView('week')\">{{ options.weekCaption }}</button>\n    <button class=\"btn btn-outline-primary\" (click)=\"toggleView('list')\">{{ options.listCaption }}</button>\n  </div>\n</div>" }]
        }], ctorParameters: function () { return []; }, propDecorators: { options: [{
                type: Input
            }], activeDate: [{
                type: Input
            }], mondayFirst: [{
                type: Input
            }], activeView: [{
                type: Input
            }], viewChange: [{
                type: Output
            }], previousBtnClick: [{
                type: Output
            }], nextBtnClick: [{
                type: Output
            }], todayBtnClick: [{
                type: Output
            }] } });
