// mdb-angular-mention/esm2020/lib/mention.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { NgModule } from '@angular/core';
import { MdbMentionItemComponent } from './mention-item.component';
import { MdbMentionComponent } from './mention.component';
import { MdbMentionDirective } from './mention.directive';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class MdbMentionModule {
}
MdbMentionModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbMentionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbMentionModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbMentionModule, declarations: [MdbMentionComponent, MdbMentionDirective, MdbMentionItemComponent], imports: [OverlayModule, CommonModule], exports: [MdbMentionComponent, MdbMentionDirective, MdbMentionItemComponent] });
MdbMentionModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbMentionModule, imports: [[OverlayModule, CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbMentionModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbMentionComponent, MdbMentionDirective, MdbMentionItemComponent],
                    imports: [OverlayModule, CommonModule],
                    exports: [MdbMentionComponent, MdbMentionDirective, MdbMentionItemComponent],
                }]
        }] });
