// mdb-angular-mention/esm2020/lib/mention-item.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class MdbMentionItemComponent {
    constructor() {
        this.optionHeight = 35;
        this.showImg = false;
        this.mentionValue = '';
        this.active = false;
    }
    setActiveStyles() {
        if (!this.active) {
            this.active = true;
        }
    }
    setInactiveStyles() {
        if (this.active) {
            this.active = false;
        }
    }
    highlightMatchingLetters(itemName) {
        const matchingLetters = new RegExp(`(?<captureGroup>${this.mentionValue})`, 'gi');
        return itemName.replace(matchingLetters, '<strong>$<captureGroup></strong>');
    }
}
MdbMentionItemComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbMentionItemComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
MdbMentionItemComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.2.7", type: MdbMentionItemComponent, selector: "mention-item", inputs: { item: "item", queryBy: "queryBy", optionHeight: "optionHeight", showImg: "showImg", mentionValue: "mentionValue" }, ngImport: i0, template: "<li #mentionItem\n    [ngStyle]=\"{'height': optionHeight + 'px'}\"\n    [ngClass]=\"{'active': active}\"\n    class=\"mention-item\">\n    <span\n    [innerHTML]=\"highlightMatchingLetters(item[queryBy])\"\n    ></span>\n    <img *ngIf=\"item.image && item.image !== '' && showImg\" [src]=\"item.image\" class=\"mention-item-image rounded-circle\">\n</li>", directives: [{ type: i1.NgStyle, selector: "[ngStyle]", inputs: ["ngStyle"] }, { type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.7", ngImport: i0, type: MdbMentionItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'mention-item', styles: [], template: "<li #mentionItem\n    [ngStyle]=\"{'height': optionHeight + 'px'}\"\n    [ngClass]=\"{'active': active}\"\n    class=\"mention-item\">\n    <span\n    [innerHTML]=\"highlightMatchingLetters(item[queryBy])\"\n    ></span>\n    <img *ngIf=\"item.image && item.image !== '' && showImg\" [src]=\"item.image\" class=\"mention-item-image rounded-circle\">\n</li>" }]
        }], ctorParameters: function () { return []; }, propDecorators: { item: [{
                type: Input
            }], queryBy: [{
                type: Input
            }], optionHeight: [{
                type: Input
            }], showImg: [{
                type: Input
            }], mentionValue: [{
                type: Input
            }] } });
