// mdb-angular-treeview/esm2020/lib/treeview.module.mjs
import { NgModule } from '@angular/core';
import { MdbTreeviewComponent } from './treeview.component';
import { MdbCheckboxModule } from 'mdb-angular-ui-kit/checkbox';
import { CommonModule } from '@angular/common';
import { MdbCollapseModule } from 'mdb-angular-ui-kit/collapse';
import * as i0 from "@angular/core";
export class MdbTreeviewModule {
}
MdbTreeviewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbTreeviewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MdbTreeviewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbTreeviewModule, declarations: [MdbTreeviewComponent], imports: [CommonModule, MdbCheckboxModule, MdbCollapseModule], exports: [MdbTreeviewComponent] });
MdbTreeviewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbTreeviewModule, imports: [[CommonModule, MdbCheckboxModule, MdbCollapseModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.3", ngImport: i0, type: MdbTreeviewModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MdbTreeviewComponent],
                    imports: [CommonModule, MdbCheckboxModule, MdbCollapseModule],
                    exports: [MdbTreeviewComponent],
                }]
        }] });
